### ISSUE TITLE
Page layout assignment through the Metadata Api is not properly working

### ISSUE THREAD

**Issue by CraigHoneyfieldIS** [financialforcedev/apex-mdapi#163]
I am using the following code to try and automate pagelayout assignment for a given profile.  It runs fine without any errors.  When I go into setup/profiles/ view assignment, it appears to be assigned to the correct page layout. 

However, when I try to open up a record an view it, it is still using the prior page layout.  If I go back into settings/profiles/ and edit/save the assignment (make no changes, just edit/save) it appears to work.  So I must be missing a step when I am setting the Page Layout Assignment?  I'm stumped.  Any help would be very much appreciated!

`  
public static void editPageLayoutAssignmentsForProfile(string profileName, List<string>pageLayoutNames) {
        if (!String.isBlank(profileName)) {
            MetadataService.MetadataPort service = createService();
            MetadataService.IReadResult readResult =
                    service.readMetadata('Profile', new String[]{
                            profileName
                    });
            MetadataService.Metadata[] records = readResult.getRecords();

            MetadataService.Profile p = (MetadataService.Profile) records[0];
            System.Debug('Found Profile:' + p.fullName);
            p.tabVisibilities = new MetadataService.ProfileTabVisibility[]{}; // this line of code is required to get around some bogus tab error:  "You can't edit tab settings for WorkRewardFund, as it's not a valid tab.

            for (String pageLayout : pageLayoutNames) {

                Boolean isClassEnabled = false;

                MetadataService.ProfileLayoutAssignment pl = new MetadataService.ProfileLayoutAssignment();

                //Remove existing layout
                Integer intIndex = 0;
                for(MetadataService.ProfileLayoutAssignment pa : p.layoutAssignments) {
                    if(pa.layout.substring(0,25) == pageLayout.substring(0,25)) {
                        p.layoutAssignments.remove(intIndex);
                        System.Debug('Removed Assignment: ' + pa.layout);
                        break;
                    }
                    intIndex++ ;
                }

                pl.layout = pageLayout;

                pl.recordType = null;

                p.layoutAssignments.add(pl);
            }

            MetadataService.SaveResult[] sr = service.updateMetadata(new MetadataService.Metadata[]{p
            });
            if (sr != null) {
                System.debug('editPageLayoutAssignmentsForProfile error ' + sr[0].errors);
            }

        }
    }
`

**Comment by afawcett** [comment]
Hmmm this could be some kind of caching thing i suspect. If you use something like Force.com or Developer Workbench to pull down the profile metadata (you will need to include your object and tab in the package manifest), does the profile xml differ before and after you resave it in the UI?

**Comment by CraigHoneyfieldIS** [comment]
Nope, the profile xml is the same before/after I resave in the UI.  How do you flush the cache?

**Comment by afawcett** [comment]
As far as i know there is no way to programatically flush the cache. It seems like this is being done by reading the profile from the UI. I wonder if doing this via SOQL might do it?

**Comment by afawcett** [comment]
@CraigHoneyfieldIS please comment further if you are still having issues here.

**Comment by jingramcamfed** [comment]
Just come across the exactly the same problem, using C# and wsdl/soap 44.0.  Only clicking Edit on a profile, then Save without any changes, gets the page layout changes available.  I found the same behaviour with Tab Settings - I set to DefaultOn rather than Hidden and the tab remains hidden until Edit/Save on the profile is carried out.

Incidentally I don't read any metadata in the program, simply pull together a new Profile object and apply it to existing profiles in Salesforce with UpdateMetadata(Metadata[]).

I have only a dozen profiles, and one database, to apply this to, so manual update/save is ok.  But as everything else I do works just fine programmatically (creating new objects, fields, page layouts, tabs) it's a niggle.
